design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,291 @@
+# KLIP-1: Improve UDF Interfaces
+
+**Author**: agavra | 
+**Release Target**: 5.3 | 
+**Status**: In Discussion | 
+**Discussion**: link
+
+**tl;dr:** *Address the most common feature requests related to UDF/UDAFs including
+struct support, generic types, variable arguments and complex aggregation*
+
+
+## Motivation and background
+
+There have been many requests ([#2163](https://github.com/confluentinc/ksql/issues/2163)) to
+improve the UDF/UDAF interface in KSQL. Of these features, four stand out:
+
+- _Custom Struct Support_ - UDFs can only operate on standard Java types (`Map`, `List`, `String`, 
+etc...) and do not accept structured types. This dramatically limits what KSQL can handle, as
```

Comment:
As of now, support for `Map` and `List` also is limited and you cannot have nested complex data.

Reply:
That's rather unfortunate 😢 but I think nested complex data is out of scope for this specific one. I'll add that to the out of scope section unless you feel strongly that those should be covered as part of this KLIP.

Reply:
might not be super-urgent compared to teh other items under discussion, but "just enough" nested support to allow e.g. a "flatten" function (`List<T> flatten(List<List<T>>`) would be useful.